tools/webdriver/webdriver/bidi/client.py

```diff
@@ -0,0 +1,393 @@
+from urllib.parse import urljoin, urlparse
+
+import websockets
+
+
+logger = logging.getLogger("webdriver.bidi")
+
+
+class BidiException(Exception):
+    def __init__(self, err: str, msg: str, stack: Optional[str] = None):
+        self.err = err
+        self.msg = msg
+        self.stack = stack
+
+
+def get_running_loop() -> asyncio.AbstractEventLoop:
+    if hasattr(asyncio, "get_running_loop"):
+        # Python 3.7+
+        return asyncio.get_running_loop()
+    # Unlike the above, this will actually create an event loop
+    # if there isn't one; hopefully running tests in Python >= 3.7
+    # will allow us to catch any behaviour difference
+    return asyncio.get_event_loop()
+
+
+class BidiSession:
+    """A WebDriver BiDi session.
+
+    This is the main representation of a BiDi session and provides the
+    interface for running commands in the session, and for attaching
+    event handlers to the session. For example:
+
+    async def on_log(data):
+        print(data)
+
+    session = BidiSession("ws://localhost:4445", capabilities)
+    session.add_event_listener("log.entryAdded", on_log)
+    await session.start()
+    await session.subscribe("log.entryAdded")
+    # Do some stuff with the session
+    session.end()
+
+    If the session id is provided it's assumed that the underlying
+    WebDriver session was already created, and the WebSocket URL was
+    taken from the new session response. If no session id is provided, it's
+    assumed that a BiDi-only session should be created when start() is called.
+
+    It can also be used as a context manager, with the WebSocket transport
+    implictly being created when the context is entered, and closed when
+    the context is exited.
+
+    :param websocket_url: WebSockets URL on which to connect to the session.
+                          This excludes any path component.
+    :param session_id: String id of existing HTTP session
+    :param capabilities: Capabilities response of existing session
+    :param requested_capabilities: Dictionary representing the capabilities request.
+
+    """
+
+    def __init__(self,
+                 websocket_url: str,
+                 session_id: Optional[str] = None,
+                 capabilities: Optional[Mapping[str, Any]] = None,
+                 requested_capabilities: Optional[Mapping[str, Any]] = None,
+                 loop: Optional[asyncio.AbstractEventLoop] = None):
+        self.transport: Optional[Transport] = None
+
+        # The full URL for a websocket looks like
+        # ws://<host>:<port>/session when we're creating a session and
+        # ws://<host>:<port>/session/<sessionid> when we're connecting to an existing session.
+        # To be user friendly, handle the case where the class was created with either a
+        # full URL including the path, and also the case where just a server url is passed in.
+        parsed_url = urlparse(websocket_url)
+        if parsed_url.path == "/":
```

Comment:
`urlparse("ws://localhost:5678").path == ""`, so what about:
```suggestion
        if parsed_url.path == "" or parsed_url.path == "/":
```